module/zfs/vdev.c

```diff
@@ -1445,8 +1459,16 @@ vdev_metaslab_group_create(vdev_t *vd)
 				spa->spa_min_ashift = vd->vdev_ashift;
 
 			uint64_t min_alloc = vdev_get_min_alloc(vd);
-			if (min_alloc < spa->spa_min_alloc)
+			if (spa->spa_min_alloc == INT_MAX &&
+			    spa->spa_gcd_alloc == INT_MAX) {
 				spa->spa_min_alloc = min_alloc;
+				spa->spa_gcd_alloc = min_alloc;
+			} else {
+				if (min_alloc < spa->spa_min_alloc)
+					spa->spa_min_alloc = min_alloc;
+				spa->spa_gcd_alloc = vdev_gcd(min_alloc,
+				    spa->spa_gcd_alloc);
+			}
```

Comment:
Why not make spa_min_alloc and spa_gcd_alloc logic completely separate.  Joined together it makes me worry about possible combinations.